parser/context.go

```diff
@@ -0,0 +1,11 @@
+package parser
+
+import (
+	"github.com/google/go-jsonnet/ast"
+	"github.com/google/go-jsonnet/internal/parser"
+)
+
+// Children returns all children of a node. It supports ASTs before and after desugaring.
+func Children(node ast.Node) []ast.Node {
```

Comment:
How about create a new dir called visitors.  I have ported the visitor from the formatter which I can put in there too.  We also have clone().

Reply:
This is a good argument for having a separate package. I am not a fan of the name `visitors`, though:
1) Not all things that we may want to put there will be strictly visitors. You mentioned Clone(), which is not implemented as a visitor (unless you understand it very broadly). Also why should the user even care if it's a visitor inside or something else. 
2) As you may remember I don't like the visitor pattern in general - IMO, most of the time it's an unnecessary straightjacket, which sometimes makes things way more difficult that they need to be. So I would prefer not to prime people to use a visitor whenever they want to do something with the AST.

What about `astutils` or `astutil`?

Also, should we allow this new package, however it is called, to depend on `jsonnet` package? This would allow putting there things like the @sh0rez's `importRecursive` there (https://github.com/grafana/tanka/pull/84/files#diff-84ccc97f3752d43b639b60a819c174f7R50).

Reply:
I also didn't like the name visitors.  astutils sounds fine.

I think we might want to use astutils internally in the VM (not just in separate executables) so that dependency might be a bad idea.

Reply:
> Putting it in ast may cause trouble with exporting directChildren etc. internally.

Can you clarify this challenge? My instinct is that if it has to do with manipulating ASTs, try to put it in the `ast` package. I'd like to understand the challenge better.

Reply:
@seh 

> Can you clarify this challenge? My instinct is that if it has to do with manipulating ASTs, try to put it in the ast package. I'd like to understand the challenge better.

1) Let's assume we do not want to expose DirectChildren in the public API. We still want to use it in other packages internally, e.g. in `linter` and in `parser`. Therefore DirectChildren may not be in the `ast` package (or it would need to be either completely public or package-private).
2) `DirectChildren` depends on `ast` package.
2) `Children` depends on `DirectChildren`. So it uses a package which depends on `ast`. So it cannot be in `ast`.

I find it quite frustrating that in Go we can make things internal only at the granularity of packages. Because of that (and a very reasonable no-circular-dependencies rule) exposing internally some part of implementation may "push out" some part of the public API.

Reply:
You could have more granular support by using a combination of an `internal` package and a public stub package:

- All code lives in the internal package, exporting everything that might be required inside of this project and for third party ones
- Because the package is `internal`, it can only be imported inside this repository
- A non-internal package with a similar name would be created that just calls the `internal` package like to:

**`/internal/mypackage/foo.go`:**
```go
// Foo might be used by third party programs
func Foo() string { return "foo" + Bar() }

// Bar is only used internally
func Bar() string { return "bar"}
```

**`/mypackage/stub.go`**:
```go
import "/internal/mypackage"

func Foo() string { return mypackage.Foo() }

// Only Foo will be exported to the outer world, while Bar is still available on its own inside this repository 
```

Reply:
Thank you for the explanation. Once again, I'm craving a diagram of the project's package structure. I'm not sure what to recommend yet.

Reply:
![godepgraph](https://user-images.githubusercontent.com/168158/65777631-4eedb780-e13c-11e9-91be-ddafa3facb28.png)

Reply:
I'm sure someone is going to ask, so:
```
go get github.com/kisielk/godepgraph
godepgraph github.com/google/go-jsonnet | dot -Tpng -o godepgraph.png
```

Reply:
Thank you, Dave.

May I ask why `Children` can't be a method on `ast.Node`?

Reply:
Ah, currently, _internal/parser_ depends on _ast_, so exposing `Children` on `ast.Node` would require folding the _internal/parser_ package (at least the _context.go_ file) into the _ast_ package. It looks as though _internal/parser_ operates atop ASTs. Figuring out a node's children is apparently a matter of _interpreting_ ASTs, which we've decided to keep separate from the definition of the AST structure itself.

Reply:
The children stuff shouldn't be in the parser at all.

Reply:
Ok so how about this
1) Really fundamental stuff for iterating over the AST can go in the ast package and isn't allowed to depend on anything else.
2) Anything that needs to resolve imports (because it's a whole program analyis, basically) lives somewhere else.  Maybe "analysis" is a better dir name, or "toolutils"?

Reply:
I agree with @sparkprime's suggestion. 

@sh0rez Could you rename the public package you created from `parser` to `toolutils`? We will merge it then. Later we may want to move it to `ast`, but that would require changing implementation of Children or exposing DirectChildren and friends and I don't want to make this change wait for that.

Reply:
Done that!